File: src/test/TestProgram.java
package test;

import com.chbachman.cminus.Run;

import java.io.File;

class TestProgram {

    private static final File testDir = new File("./build/test/");
    private static final String tempFileName = "temp.c";
    private static final String tempFileName2 = "temp2";
    private static final File tempFile = new File(testDir, tempFileName);
    private static final File testFileDir = new File("./src/test/cmSource");

    private static boolean test(File program) {
        try {
            String programName = program.getCanonicalPath();
            String tempName = tempFile.getCanonicalPath();
            String temp2Name = new File(testDir, tempFileName2).getCanonicalPath();

            Run.buildCM(programName, tempName);
            Run.build(tempName, temp2Name).waitFor();

            String c = Run.run(temp2Name);
            String swift = Run.command("swift " + programName);

            return c.equals(swift);
        } catch (Exception e) {
            e.printStackTrace();
            // Compilation failed.
            return false;
        }
    }

    static boolean test(String path) {
        return test(new File(testFileDir, path));
    }

}

File: src/test/java/TestProgram.kt
import com.chbachman.cminus.util.Output
import com.chbachman.cminus.util.Run
import org.testng.Assert.assertEquals
import org.testng.Assert.fail
import java.io.File
import java.io.IOException

internal object TestProgram {
    private val testFileDir = File("./src/test/cm")

    private fun test(program: File, outputTest: String? = null) {
        val tempDir = createTempDir()
        tempDir.deleteOnExit()

        val fileName = program.nameWithoutExtension

        val cFile = File(tempDir, "$fileName.c")
        val executable = File(tempDir, fileName)
        val kotlinScript = File(tempDir, "$fileName.kts")

        try {
            Run.buildCM(program, cFile)
            Run.build(cFile, executable).print()

            val c = Run.run(executable)

            // If we were not given any output to check against, check it against kotlin itself.
            val programOutput =
                if (outputTest == null) {
                    program.copyTo(kotlinScript, true)
                    kotlinScript.appendText("\nmain()")

                    Run.command("kotlinc -nowarn -script ${kotlinScript.canonicalPath}")
                } else {
                    Output(outputTest, "")
                }

            if (c != programOutput) {
                println("C Minus Output:")
                println(c.out)

                println("Kotlin Output:")
                println(programOutput.out)
            }

            assertEquals(c, programOutput)
        } catch (err: IOException) {
            val files = tempDir.listFiles().toList()
            print(files)

            files.forEach {
                it.readLines().forEach {
                    println(it)
                }
            }

            throw err
        }
    }

    fun testSimple(program: String, output: String) {
        val temp = createTempFile()
        temp.writeText(program)

        try {
            test(temp, output)
        } catch (err: IOException) {
            fail("Could not recover from:", err)
        }
    }

    fun test(path: String) {
        try {
            test(File(testFileDir, path))
        } catch (err: IOException) {
            fail("Could not recover from:", err)
        }
    }
}
